### PR TITLE
callback and update qth

### PR DESCRIPTION
**Pull request by rainarashika**

### CODE REVIEW

File: dqn_lstm.py [hunk 1]
```diff
@@ -93,6 +137,15 @@ def replay(self, batch_size):
     def predict(self, *args, **kwargs):
         # Directly call the predict method of the encapsulated Keras model
         return self.model.predict(*args, **kwargs)
+    
+    def compute_E_Q_less_qth(self):
+        #Computes E[Q | Q < qth] using the current model's predictions.
+        X, _ = OutageData(**self.data_config).__getitem__(0)
```
Comment (desimmons) [inline dqn_lstm.py:143]: You need to perform this in more than just one item, I think

File: dqn_lstm.py [hunk 2]
```diff
@@ -12,13 +12,56 @@
...
+        self.data_config = data_config
+        self.step_size = step_size
+        self.threshold = threshold
+        
+
+    def on_epoch_end(self, epoch, logs=None):
+        X, y_true = OutageData(**self.data_config).__getitem__(0)
+        y_pred = self.model_wrapper.model.predict(X)  
+
+        P_infty_estimate = self.model_wrapper.model.loss(y_true, y_pred)
```
Comment (desimmons) [inline dqn_lstm.py:43]: Same with this. You need to loop over many page examples to form the estimate. One `__item__` call isn't going to be statistically significant

File: dqn_lstm.py [hunk 3]
```diff
@@ -53,22 +102,25 @@ def build_model(self):
...
                                     tf.keras.metrics.BinaryAccuracy(),
                                     tf.keras.metrics.Accuracy()])
             else:
                 raise ValueError(f"Invalid loss name: {self.model_name}")
             print(f"Training model: {self.model_name}")
+            self.model = model
             training_generator = OutageData(**self.data_config)
-            callback = tf.keras.callbacks.EarlyStopping(monitor='loss', patience=self.epochs, restore_best_weights=True)
-            history = model.fit(training_generator, epochs=self.epochs, callbacks=[callback])
+            qth_callback = QthUpdateCallback(self,step_size=0.01, threshold=1e-3)
```
Comment (desimmons) [inline dqn_lstm.py:114]: I imagine what's being done here is causing issues. IIUC, this bit of code is being called inside the constructor of `self`, so you're passing a loss function `self`, which hasn't been fully constructed, to the callback object, which, inside its own constructor, takes a copy of the unconstructed loss function and then later tries to use it

Comment (desimmons) [inline dqn_lstm.py:114]: I think you need to define a variable `loss=Finite...` above all of this and pass it to both `model.compile...` and `QthUpdate...`

**Review by desimmons — CHANGES_REQUESTED** [review]
